File: algorithm/_misc/binary-search-tree-and-sorted-double-linked-list.cpp
// Binary Search Tree and Sorted Double Linked List


// Sorted Double Linked List to Binary Tree
class Solution {
public:
  DoubleListNode* sortedDoubleList2BST(DoubleListNode* head) {
    int count = 0;
    for (DoubleListNode* curr = head; curr; curr = curr->next) {
      ++count;
    }
    return sortedDoubleList2BST(head, count);
  }

private:
  DoubleListNode* sortedDoubleList2BST(DoubleListNode*& head, int n) {
    if (n <= 0) {
      return nullptr;
    }
    DoubleListNode* left = sortedDoubleList2BST(head, n / 2);

    DoubleListNode* root = head;
    root->prev = left;

    head = head->next;
    root->next = sortedDoubleList2BST(head, n - n / 2 - 1);

    return root;
  }
};

// Binary Tree to Sorted Double Linked List
class Solution {
public:
  void BST2sortedDoubleList(TreeNode* root, TreeNode*& head) {
    TreeNode* prev = nullptr;
    BST2sortedDoubleList(root, head, prev)
  }

private:
  void BST2sortedDoubleList(TreeNode* root, TreeNode*& head, TreeNode*& prev) {
    if (!root) {
      return;
    }
    BST2sortedDoubleList(root.left, head, prev);

    if (prev == nullptr) {
      head = root;
    } else {
      root.left = prev;
      prev.right = root;
    }
    prev = root;

    BST2sortedDoubleList(root.right, head, prev);
  }
};
  

File: data-structure-and-algorithm/typescript/leetcode/src/extra-questions/binary-search-tree-and-sorted-double-linked-list.ts
/*
Binary Search Tree and Sorted Double Linked List
*/

import {PerformanceObserver} from 'perf_hooks';
import {TreeNode} from '../data-structure/BinaryTree';
import {DoubleListNode} from '../data-structure/DoublyLinkedList';
import {Nullable} from '../util/object';

function sortedDoubleList2BST(
  head: Nullable<DoubleListNode<number>>
): Nullable<DoubleListNode<number>> {
  let cnt = 0;
  for (let curr = head; curr; curr = curr.next) {
    ++cnt;
  }
  return sortedDoubleList2BSTDfs(cnt);

  function sortedDoubleList2BSTDfs(
    size: number
  ): Nullable<DoubleListNode<number>> {
    if (size <= 0) {
      return null;
    }

    const left = sortedDoubleList2BSTDfs(Math.trunc(size / 2));
    const node = head!;
    head = head!.next;
    node.prev = left;
    node.next = sortedDoubleList2BSTDfs(size - Math.trunc(size / 2) - 1);

    return node;
  }
}

function BST2sortedDoubleList(
  root: Nullable<TreeNode<number>>
): Nullable<TreeNode<number>> {
  if (root === null) {
    return null;
  }
  let prev: Nullable<TreeNode<number>> = null;
  let head: Nullable<TreeNode<number>> = null;

  BST2sortedDoubleListDfs(root);

  prev!.right = head;
  head!.left = prev;

  return head;

  function BST2sortedDoubleListDfs(node: Nullable<TreeNode<number>>): void {
    if (node === null) {
      return;
    }
    BST2sortedDoubleListDfs(node.left);

    if (head === null) {
      head = node;
    } else {
      node.left = prev;
      prev!.right = node;
    }
    prev = node;

    BST2sortedDoubleListDfs(node.right);
  }
}
